[ZEPPELIN-1046] bin/install-interpreter.sh for netinst package

Leemoonsoo opened the pull request:
### What is this PR for?
Implementation of bin/install-interpreter.sh for netinst package which suggested in the [discussion](http://apache-zeppelin-users-incubating-mailing-list.75479.x6.nabble.com/Ask-opinion-regarding-0-6-0-release-package-tp3298p3314.html).

Some usages will be

```
# download all interpreters provided by Apache Zeppelin project
bin/install-interpreter.sh --all

# download an interpreter with name (for example markdown interpreter)
bin/install-interpreter.sh --name md

# download an (3rd party) interpreter with specific maven artifact name
bin/install-interpreter.sh --name md -t org.apache.zeppelin:zeppelin-markdown:0.6.0-SNAPSHOT
```

If it looks fine, i'll continue the work (refactor code, and add test)

### What type of PR is it?
Feature

### Todos
* [x] - working implementation
* [x] - refactor
* [x] - add test

### What is the Jira issue?
* Open an issue on Jira https://issues.apache.org/jira/browse/ZEPPELIN/
* Put link here, and add [ZEPPELIN-*Jira number*] in PR title, eg. [ZEPPELIN-533]

### How should this be tested?
Outline the steps to test the PR here.

### Screenshots (if appropriate)

### Questions:
* Does the licenses files need update?
* Is there breaking changes for older versions?
* Does this needs documentation?

prabhjyotsingh commented:
One more feature request for this, with downloading an interpreter, creation/insertion of relevant entry in `conf/interpreter.json`

felixcheung commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/install/InstallInterpreter.java:
```diff
@@ -0,0 +1,161 @@
+    "md",     "zeppelin-markdown",   "Markdown interpreter",
+    "sh",     "zeppelin-shell",      "Allows shell command",
+    "jdbc",   "zeppelin-jdbc",
+    "Generic JDBC interpreter for hive, phoenix, postsgresql, mysql, etc",
```
do you think we would need to update this list in the future?
is there a way to have a text lookup table with a list of short name instead of hardcoding it in .java file?

bzz commented:
Both suggestions are nice addition, and there is also a refactoring on TODOs so I guess it's WIP now.

👍  for making it part of the release, as discussed dev@z.a.o

Leemoonsoo commented:
@prabhjyotsingh I think that's good idea.

If this script runs before start Zeppelin, right after download and unarchive, Zeppelin will auto create conf/interpreter.json with all installed interpreters. 

If this script runs after conf/interpreter.json created, it'll not update conf/interpreter.json. And i think it make sense while conf/interpreter.json might customized by user at this time, and user may not want to tough/add something into it.

I think we'll have better way to install interpreter once Zeppelin have dynamic interpreter loading capability (https://issues.apache.org/jira/browse/ZEPPELIN-598)

@felixcheung I have removed hardcoded list and updated to read list from text file `conf/interpreter-list`.

minahlee commented on zeppelin-interpreter/src/main/java/org/apache/zeppelin/dep/DependencyResolver.java:
```diff
@@ -113,6 +113,28 @@ public DependencyResolver(String localRepoPath) {
     return libs;
   }
 
+  public List<File> load(String artifact, File destPath) throws IOException, RepositoryException {
```
Seems like this method with same signature already exists [above]( https://github.com/Leemoonsoo/zeppelin/blob/755912c72458326a25bb26f9bf249a4a4bbf3b1b/zeppelin-interpreter/src/main/java/org/apache/zeppelin/dep/DependencyResolver.java#L66).

minahlee commented on zeppelin-interpreter/src/main/java/org/apache/zeppelin/dep/DependencyResolver.java:
```diff
@@ -113,6 +113,28 @@ public DependencyResolver(String localRepoPath) {
+    return load(artifact, new LinkedList<String>(), destPath);
+  }
+
+  public List<File> load(String artifact, Collection<String> excludes, File destPath)
```
Same here

AhyoungRyu commented on conf/interpreter-list:
```diff
@@ -0,0 +1,35 @@
+alluxio         org.apache.zeppelin.zeppelin-alluxio:0.6.0              Alluxio interpreter
+angular         org.apache.zeppelin.zeppelin-angular:0.6.0              HTML and AngularJS view rendering
+cassandra       org.apache.zeppelin.zeppelin-cassandra:0.6.0            Cassandra interpreter
+elasticsearch   org.apache.zeppelin.zeppelin-cassandra:0.6.0            Elasticsearch interpreter
```
@Leemoonsoo Do you mean  `org.apache.zeppelin.zeppelin-elasticsearch:0.6.0` ?

Leemoonsoo commented:
It's ready to review and merge.
Tried updated create_release.sh script. `zeppelin-0.6.0-bin-all.tgz` is about 498M while `zeppelin-0.6.0-netinst.tgz` will be about 208MB.

minahlee commented:
@Leemoonsoo Could you rebase and resolve conflicts?

minahlee commented on docs/manual/interpreterinstallation.md:
```diff
@@ -0,0 +1,164 @@
+  <tr>
+    <td>cassandra</td>
+    <td>org.apache.zeppelin.zeppelin-cassandra:0.6.0</td>
+    <td>HTML and AngularJS view rendering</td>
```
Seems like description needs to be updated

minahlee commented on docs/manual/interpreterinstallation.md:
```diff
@@ -0,0 +1,164 @@
+
+The above command will download maven artifact `groupId1:artifact1:version1` and all of it's transitive dependencies into `interpreter/interpreter1` directory.
+
+Once you have installed interpreters, you'll need to add interpreter class name into `zeppelin.interpreters` property in [configuration](../install/install.html#zeppelin-configuration).
```
`zeppelin-configuration` -> `apache-zeppelin-configuration`

minahlee commented on docs/_includes/themes/zeppelin/_navigation.html:
```diff
@@ -21,7 +21,7 @@
                 <li class="title"><span><b>Getting Started</b><span></li>
-                <li><a href="{{BASE_PATH}}/install/install.html">Quick Start</a></li>
+                <li><a href="{{BASE_PATH}}/install/install.html">Install</a></li>
                 <li><a href="{{BASE_PATH}}/install/install.html#zeppelin-configuration">Configuration</a></li>
```
`zeppelin-configuration` -> `apache-zeppelin-configuration`

minahlee commented on conf/interpreter-list:
```diff
@@ -0,0 +1,35 @@
+#
+# [name]  [maven artifact]  [description]
+
+alluxio         org.apache.zeppelin.zeppelin-alluxio:0.6.0              Alluxio interpreter
```
After org.apache.zeppelin, `:` instead of `.`

Leemoonsoo commented:
@minahlee addressed your comments

minahlee commented on docs/manual/interpreterinstallation.md:
````diff
@@ -0,0 +1,164 @@
+#### Install 3rd party interpreters
+
+```
+./bin/install-interpreter.sh --name interpreter1 --repository groupId1:artifact1:version1
````
`--repository` -> `--artifact`

minahlee commented on docs/manual/interpreterinstallation.md:
````diff
@@ -0,0 +1,164 @@
+#### Install multiple 3rd party interpreters at once
+
+```
+./bin/install-interpreter.sh --name interpreter1,interpreter2 --repository groupId1:artifact1:version1,groupId2:artifact2:version2
````
`--repository` -> `--artifact`

minahlee commented on docs/manual/interpreterinstallation.md:
````diff
@@ -0,0 +1,164 @@
+./bin/install-interpreter.sh --name interpreter1,interpreter2 --repository groupId1:artifact1:version1,groupId2:artifact2:version2
+```
+
+`--name` and `--repository` arguments will recieve comma separated list.
````
`--repository` -> `--artifact`

minahlee commented on dev/create_release.sh:
```diff
@@ -103,6 +103,7 @@ function make_binary_release() {
 git_clone
 make_source_package
 make_binary_release all "-Pspark-1.6 -Phadoop-2.4 -Pyarn -Ppyspark -Psparkr -Pr"
+make_binary_release netinst "-Pspark-1.6 -Phadoop-2.4 -Pyarn -Ppyspark -pl '!alluxio,!angular,!cassandra,!elasticsearch,!file,!flink,!hbase,!ignite,!jdbc,!kylin,!lens,!livy,!markdown,!postgresql,!python,!shell"
```
Single quote in front of `!alluxio` should be removed here

minahlee commented:
Thanks for the effort @Leemoonsoo, @AhyoungRyu I've tested this branch and it works like a charm! I left few more comments, once they are addressed it's good to merge :)

Leemoonsoo commented:
Thanks @minahlee for the review. Addressed your comments!

minahlee commented:
Merging if there is no more discussion
